### ISSUE TITLE
ExecuteUpdate on Nullable stop working after migration from 8 to 9

### ISSUE THREAD

**Issue by vzabavnov** [dotnet/efcore#35164]
the test:

```c#
public enum Status { Good, NotGood }

public class Entity
{
    public int ID { get; set; }

    public string Value { get; set; } = null!;

    public Status? Status { get; set; }
}

public class Context(DbContextOptions<Context> options) : DbContext(options)
{
    public DbSet<Entity> Entities { get; set; }

    protected override void OnModelCreating(ModelBuilder builder)
    {
        base.OnModelCreating(builder);
        
        builder.Entity<Entity>(entity =>
        {
            entity.HasKey(z => z.ID);
            entity.HasData(new Entity() { ID = 1, Value = "Value 1" });
        });
    }
}

public class UnitTest
{
    [Fact]
    public void Test()
    {
        var connection = new SqliteConnection("datasource=:memory:");
        connection.Open();
        var options = new DbContextOptionsBuilder<Context>().UseSqlite(connection).Options;

        using (var ctx = new Context(options))
        {
            ctx.Database.EnsureCreated();
            
            Assert.Single(ctx.Entities);
            Assert.Contains(ctx.Entities, z => z.Value == "Value 1");
            Assert.DoesNotContain(ctx.Entities, z => z.Status != null);
        }

        using (var ctx = new Context(options))
        {
            ctx.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => Status.Good));
        }
        
        using (var ctx = new Context(options))
        {
            Assert.True(ctx.Entities.All(z => z.Status == Status.Good));
        }
    }
}
```
Works fine in all 8s versions.

In 9.0 it gives exception:
```
System.InvalidOperationException
No coercion operator is defined between types 'System.Func`2[EfCore9Bug.Entity,System.Nullable`1[EfCore9Bug.Status]]' and 'EfCore9Bug.Status'.
```
The work around is declare variable with ***Status*** outside of the ExecuteUpdate and cast it to ***Nullable***:
<pre><code>
using (var ctx = new Context(options))
{
    <b>var status = (Status?)Status.Good;</b>
    ctx.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => <b>status)</b>);
}
</code></pre>

**Comment by roji** [comment]
@vzabavnov I can indeed see the error happening, but I'm getting the same error in 8.0 with the code below (which is a simplified version of your code above). Can you please double-check and confirm whether you're seeing the exception on EF 8.0 or not?

<details>
<summary>Repro</summary>

```c#
await using var context = new MyContext();
await context.Database.EnsureDeletedAsync();
await context.Database.EnsureCreatedAsync();

// Throws "No coercion operator" on both EF 8.0 and 9.0
context.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => Status.Good));

// Works on 8.0, on 9.0 throws "unbound variable: __status_0" (this is #35152, already fixed)
// var status = Status.Good;
// context.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => status));

public class MyContext : DbContext
{
    public DbSet<Entity> Entities { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        => optionsBuilder
            // .UseSqlServer("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false")
            .UseSqlite("datasource=/tmp/foo.sqlite")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
}

public class Entity
{
    public int Id { get; set; }
    public string Value { get; set; } = null!;
    public Status? Status { get; set; }
}

public enum Status { Good, NotGood }
```

</details>

**Comment by vzabavnov** [comment]
Just checked again - ***works***. here is my proj file:
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <!--<TargetFramework>net9.0</TargetFramework>-->
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
    <IsPackable>false</IsPackable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="coverlet.collector" Version="6.0.2" />
    
    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.11.1" />
    <PackageReference Include="xunit" Version="2.9.2" />
    <PackageReference Include="xunit.runner.visualstudio" Version="2.8.2" />
    <!--<PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="9.0.0" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Relational" Version="9.0.0" />-->
    <PackageReference Include="Microsoft.EntityFrameworkCore" Version="8.0.11" />
    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.11" />
  </ItemGroup>

  <ItemGroup>
    <Using Include="Xunit" />
  </ItemGroup>

</Project>

```

**Comment by roji** [comment]
Thanks, I can see it now (and thanks for the minimal repro)! Will investigate.

<details>
<summary>Slightly simplified repro</summary>

```c#
using var ctx = new Context();
ctx.Database.EnsureDeleted();
ctx.Database.EnsureCreated();

//var status = (Status?)Status.Good;
//ctx.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => status));
ctx.Entities.ExecuteUpdate(u => u.SetProperty(z => z.Status, _ => Status.Good));

public class Context : DbContext
{
    public DbSet<Entity> Entities { get; set; }

    protected override void OnConfiguring(DbContextOptionsBuilder builder)
    {
        builder
            .UseSqlServer("Server=localhost;Database=test;User=SA;Password=Abcd5678;Connect Timeout=60;ConnectRetryCount=0;Encrypt=false")
            .LogTo(Console.WriteLine, LogLevel.Information)
            .EnableSensitiveDataLogging();
    }
}

public class Entity
{
    public int Id { get; set; }
    public Status? Status { get; set; }
}

public enum Status { Good, NotGood }
```

</details>

**Comment by roji** [comment]
Duplicate of #35152

**Comment by roji** [comment]
The root cause here is the same as the one for #35152: the funcletizer doesn't handle lambda evaluatability correctly. I've confirmed that testing against #35172 (the PR that resolves #35152) passes while testing against the commit just before fails.
